src/dscanner/reports.d

```diff
@@ -0,0 +1,160 @@
+//			Copyright Brian Schott (Hackerpilot) 2012.
+// Distributed under the Boost Software License, Version 1.0.
+//	  (See accompanying file LICENSE_1_0.txt or copy at
+//			http://www.boost.org/LICENSE_1_0.txt)
+
+module dscanner.reports;
+
+import std.json;
+import std.algorithm : map;
+import std.array : split, array, Appender;
+
+import dscanner.analysis.base : Message, MessageSet;
+
+class SonarQubeGenericIssueDataReporter
+{
+	enum Type
+	{
+		bug = "BUG",
+		vulnerability = "VULNERABILITY",
+		codeSmell = "CODE_SMELL"
+	}
+
+	enum Severity
+	{
+		blocker = "BLOCKER",
+		critical = "CRITICAL",
+		major = "MAJOR",
+		minor = "MINOR",
+		info = "INFO"
+	}
+
+	struct Issue
+	{
+		string engineId;
+		string ruleId;
+		Location primaryLocation;
+		string type;
+		string severity;
+		int effortMinutes;
+		Location[] secondaryLocations;
+	}
+
+	struct Location
+	{
+		string message;
+		string filePath;
+		TextRange textRange;
+	}
+
+	struct TextRange
+	{
+		// SonarQube Generic Issue only allows specifying start line only
+		// or the complete range, for which start and end has to differ
+		// D-Scanner does not provide the complete range info
+		long startLine;
+	}
+
+	private Appender!(Issue[]) _issues;
```

Comment:
I think this is missing initialization? `_issues = appender!(Issue[]);`

I don't think it's specified what will happen if you don't initialize it, but `appender` is documented to be a convenience function returning an instance.

The constructor of Appender is doing a bit of work so I think it should be called in the constructor in the class: https://github.com/dlang/phobos/blob/master/std/array.d#L3239

Reply:
fixed